### ISSUE TITLE
test_openssl_engine_cipher_rc4 - use a newer cipher?

### ISSUE THREAD

**Issue by MSP-Greg** [ruby/openssl#154]
Seem to be having some issues with 2.10 and 1.1.0f passing this test.  In some environments it does, others it does not.  All other tests pass.

Hence, could [test_openssl_engine_cipher_rc4](https://github.com/ruby/openssl/blob/51ff8162bd454169010eadb8f2ef727402df04db/test/test_engine.rb#L54-L74) be updated to use a newer (or more secure) cipher.

A comment in the test is:
```ruby
algo = "RC4" #AES is not supported by openssl Engine (<=1.0.0e)
```

I looked thru the travis and appveyor files, and I didn't see any testing on versions as old as 1.0.0e...

Thanks,
Greg

**Comment by rhenium** [comment]
The comment is inaccurate -- the 'openssl' engine is for testing purpose and still supports only 'rc4' and 'rc4-40' in current versions of OpenSSL. The test case should be skipped if the OpenSSL library is built without RC4 (i.e., with 'no-rc4' option).

Does modifying the test like this fix the failure?

```patch
diff --git a/test/test_engine.rb b/test/test_engine.rb
index 63435046501e..1088280b5789 100644
--- a/test/test_engine.rb
+++ b/test/test_engine.rb
@@ -52,10 +52,16 @@ def test_openssl_engine_digest_sha1
   end
 
   def test_openssl_engine_cipher_rc4
+    begin
+      OpenSSL::Cipher.new("rc4")
+    rescue OpenSSL::Cipher::CipherError
+      pend "RC4 is not supported"
+    end
+
     with_openssl <<-'end;'
       begin
         engine = get_engine
-        algo = "RC4" #AES is not supported by openssl Engine (<=1.0.0e)
+        algo = "RC4"
         data = "a" * 1000
         key = OpenSSL::Random.random_bytes(16)
         # suppress message from openssl Engine's RC4 cipher [ruby-core:41026]
```

**Comment by MSP-Greg** [comment]
> Does modifying the test like this fix the failure?

I haven't patched the test, but `rc4 = OpenSSL::Cipher.new("rc4")` seems to be fine (rc4's class is `OpenSSL::Cipher`).

This is driving me crazy...  I've got trunk building on Appveyor, and the only error is this, running in 'make test-all'.  But, just using runner, all is fine.  See [appveyor-testing](https://ci.appveyor.com/project/MSP-Greg/appveyor-testing). note line 54.

Of course, on my home system, 'make test-all' passes.

If you have any ideas, I'd greatly appreciate any help you have time for.

Greg

**Comment by MSP-Greg** [comment]
@rhenium 

I forgot the main thing you might want to see; here's the error test from the failure:
```
 86) Failure:
OpenSSL::TestEngine#test_openssl_engine_cipher_rc4 [C:/projects/ruby-loco/src/ruby/test/openssl/test_engine.rb:80]:
assert_separately failed with error message
pid 1064 exit 0
| (TEST_ENG_OPENSSL_RC4) test_init_key() called
| (TEST_ENG_OPENSSL_RC4) test_init_key() called
```

**Comment by rhenium** [comment]
Hmm, I don't know why it's failing to suppress stderr, but it should have been using ignore_stderr option of assert_separately instead.

```patch
diff --git a/test/test_engine.rb b/test/test_engine.rb
index 63435046501e..ee1ebb33eddd 100644
--- a/test/test_engine.rb
+++ b/test/test_engine.rb
@@ -52,32 +52,22 @@ def test_openssl_engine_digest_sha1
   end
 
   def test_openssl_engine_cipher_rc4
-    with_openssl <<-'end;'
-      begin
-        engine = get_engine
-        algo = "RC4" #AES is not supported by openssl Engine (<=1.0.0e)
-        data = "a" * 1000
-        key = OpenSSL::Random.random_bytes(16)
-        # suppress message from openssl Engine's RC4 cipher [ruby-core:41026]
-        err_back = $stderr.dup
-        $stderr.reopen(IO::NULL)
-        encrypted = crypt_data(data, key, :encrypt) { engine.cipher(algo) }
-        decrypted = crypt_data(encrypted, key, :decrypt) { OpenSSL::Cipher.new(algo) }
-        assert_equal(data, decrypted)
-      ensure
-        if err_back
-          $stderr.reopen(err_back)
-          err_back.close
-        end
-      end
+    with_openssl(<<-'end;', ignore_stderr: true)
+      engine = get_engine
+      algo = "RC4" #AES is not supported by openssl Engine (<=1.0.0e)
+      data = "a" * 1000
+      key = OpenSSL::Random.random_bytes(16)
+      encrypted = crypt_data(data, key, :encrypt) { engine.cipher(algo) }
+      decrypted = crypt_data(encrypted, key, :decrypt) { OpenSSL::Cipher.new(algo) }
+      assert_equal(data, decrypted)
     end;
   end
 
   private
 
   # this is required because OpenSSL::Engine methods change global state
-  def with_openssl(code)
-    assert_separately([{ "OSSL_MDEBUG" => nil }, "-ropenssl"], <<~"end;")
+  def with_openssl(code, **opts)
+    assert_separately([{ "OSSL_MDEBUG" => nil }, "-ropenssl"], <<~"end;", **opts)
       require #{__FILE__.dump}
       include OpenSSL::TestEngine::Utils
       #{code}
```

**Comment by MSP-Greg** [comment]
Sorry for the delay; since I only has issues with make test-all, I ran another ruby build.
```
16878 tests, 2236359 assertions, 0 failures, 0 errors, 117 skips
ruby -v: ruby 2.5.0dev (2017-09-24 trunk 60006) [x64-mingw32]
```

**Thank you**.

Appveyor build is now without failures or errors.  [Test results](https://ci.appveyor.com/project/MSP-Greg/ruby-loco/build/7#L2861) all the way at the bottom...

After digging around OpenSSL, I think most MinGW builds have the engine dll's in the wrong place, so I've got to adjust some code (and let the RI2 group know).  I've also got a few other build issues to sort out.
